src/webapi/events/drag.rs

```diff
@@ -0,0 +1,697 @@
+        );
+    }
+
+    /// Set the image to be used for dragging if a custom one is desired.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/setDragImage)
+    pub fn set_drag_image( &self, img: &ImageElement, x_offset: u32, y_offset: u32 ) {
+        js!(@(no_return)
+            @{self.as_ref()}.setDragImage(@{img.as_ref()}, @{x_offset}, @{y_offset});
+        );
+    }
+}
+
+/// A DOMString representing the drag operation effect.
+// https://www.w3.org/TR/html51/editing.html#dom-datatransfer-dropeffect
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum DropEffect {
+    /// A copy of the source item is made at the new location
+    Copy,
+    /// An item is moved to a new location.
+    Move,
+    /// A link is established to the source at the new location.
+    Link,
+    /// The item may not be dropped.
+    None,
+}
+
+/// A DOMString representing the drag operation that is allowed.
+// https://www.w3.org/TR/html51/editing.html#dom-datatransfer-effectallowed
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum EffectAllowed {
+    /// The item may not be dropped.
+    None,
+    /// A copy of the source item may be made at the new location.
+    Copy,
+    /// A copy or link operation is permitted.
+    CopyLink,
+    /// A copy or move operation is permitted.
+    CopyMove,
+    /// A link may be established to the source at the new location.
+    Link,
+    /// A link or move operation is permitted.
+    LinkMove,
+    /// An item may be moved to a new location.
+    Move,
+    /// All operations are permitted.
+    All,
+    /// The default value when the effect has not been set, equivalent to all.
+    Uninitialized,
+}
+
+/// The DataTransferItemList object is a list of DataTransferItem objects representing items being
+/// dragged.
+/// During a drag operation, each DragEvent has a dataTransfer property and that property is a
+/// DataTransferItemList.
+///
+/// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList)
+// https://www.w3.org/TR/html51/editing.html#the-datatransferitemlist-interface
+#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
+#[reference(instance_of = "DataTransferItemList")]
+pub struct DataTransferItemList( Reference );
+impl DataTransferItemList {
+    /// An unsigned long that is the number of drag items in the list.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList/length)
+    pub fn len( &self ) -> u64 {
+        js!(
+            return @{self.as_ref()}.length;
+        ).try_into().unwrap()
+    }
+
+    /// Adds an item (either a File object or a string) to the drag item list and returns
+    /// a DataTransferItem object for the new item.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList/add)
+    pub fn add( &self, item: DataTransferItemAdd ) -> Option<DataTransferItem> {
```

Comment:
1. Get rid of the enum and split this into two methods - `add` and `add_file` perhaps?
2. This can throw `NotSupportedError` in case of the string variant, so it should return `Result< Option< DataTransferItem >, NotSupportedError >`.

Reply:
1. I will split this into `add_string` and `add_file`

Reply:
We could also replace the type `DataTransferItem` with two types `DataTransferItemString` and `DataTransferItemFile`, for better safety. I'll let you decide what the right trade-off is between convenience and spec accuracy.

Reply:
It's generally simpler and more Rust-y to just have two methods instead of emulating overloading with enums, (And it's one less enum to `use`.) so please make it into two methods. (:

Reply:
I did already (`add_string` and `add_file`), see the current version of the PR

Reply:
What I meant, is that since we have two methods, we can have two return types encoding the two different behaviors. By default, I'll keep one return type.

Reply:
Ah, right, sorry! I replied before looking at the code. (:

Hmm... yes, you're right, we could have two different return types, although it's probably not worth it in this case.